pyquil/kraus.py

```diff
@@ -56,3 +58,211 @@ def _create_kraus_pragmas(name, qubit_indices, kraus_ops):
+        assert len(instruction.qubits) == 1
+        param_str = format_parameter(instruction.params[0])
+        if param_str == 'pi/2':
+            return Gate('noisy-x-plus90', [], instruction.qubits)
+        if param_str == '-pi/2':
+            return Gate('noisy-x-minus90', [], instruction.qubits)
+        raise ValueError("Can't add noise to a parametric gate. "
+                         "Try compiling to RX(pi/2) or RX(-pi/2)")
+
+    if instruction.name == 'CZ':
+        return Gate('noisy-cz', [], instruction.qubits)
+
+    raise ValueError('Gate {} is not in the native instruction set'.format(instruction))
+
+
+def _get_program_topology(prog):
+    """
+    Get the graph of two-qubit gates used in a program.
+
+    :param prog: The program
+    :return: (qubits, edges). ``qubits`` is a set of all integer qubits.
+        ``edges`` is a set of tuples of two-qubit pairings.
+    """
+    qubits = set()
+    edges = set()
+    for instruction in prog:
+        if isinstance(instruction, Pragma):
+            continue
+        qb = [q.index for q in instruction.qubits]
+        qubits.update(qb)
+        if len(qb) == 2:
+            edges.add(tuple(sorted(qb)))
+        if len(qb) > 2:
+            raise ValueError("Encountered a >2 qubit instruction")
+
+    return qubits, edges
+
+
+def add_noise_to_program(prog, T1=30e-6, T2=None, gate_time_1q=50e-9, gate_time_2q=150e-09):
+    """
+    Add generic damping and dephasing noise to a program.
+
+    This high-level function is provided as a convenience to investigate the effects of a
+    generic noise model on a program. For more fine-grained control, please investigate
+    the other methods available in the ``pyquil.kraus`` module.
+
+    In an attempt to closely model the QPU, noisy versions of RX(+-pi/2) and CZ are provided;
+    I and parametric RZ are noiseless, and other gates are not allowed. To use this function,
+    you need to compile your program to this native gate set.
+
+    The default noise parameters
+
+        T1 = 30 us
+        T2 = T1 / 2
+        1q gate time = 50 ns
+        2q gate time = 150 ns
+
+    are currently typical for near-term devices.
+
+    This function will define new gates and add Kraus noise to these gates. It will translate
+    the input program to use the noisy version of the gates.
+
+    :param prog: A pyquil program consisting of I, RZ, CZ, and RX(+-pi/2) instructions
+    :param T1: The T1 amplitude damping time. By default, this is 30 us
+    :param T2: The T2 dephasing time. By default, this is one-half of the T1 time.
+    :param gate_time_1q: The duration of the one-qubit gates, namely RX(+pi/2) and RX(-pi/2).
+        By default, this is 50 ns.
+    :param gate_time_2q: The duration of the two-qubit gates, namely CZ.
+        By default, this is 150 ns.
+    :return: A new program with noisy operators.
+    """
+
+    if T2 is None:
+        T2 = T1 / 2.0
+
+    from pyquil.quil import Program  # Avoid circular dependency
```

Comment:
Where does pyquil.quil call kraus.py?

Reply:
`Program` has an `add_noisy_gate` argument that calls the private methods in `kraus.py` for checking stuff

Reply:
If this circular import avoidance is too brittle, we can also move those private methods from kraus.py to a new private module, e.g., `_kraus_pragmas.py` or something like that.